smart_campus/app/views.py

```diff
@@ -1036,3 +1103,128 @@ def travelplan_delete_page(request, pk):
     travelplan.delete()
 
     return HttpResponseRedirect('/travelplans/')
+
+
+@login_required
+def question_list_page(request):
+    if request.user.is_administrator():
+        stations = Station.objects.all()
+        question_list = Question.objects.all()
+    else:
+        stations = Satation.objects.filter(owner_group=request.user.group)
+        question_list = Question.objects.filter(linked_station__in=stations)
+
+    paginator = Paginator(question_list, 10)
+
+    page = request.GET.get('page', 1)
+    try:
+        questions = paginator.page(page)
+    except PageNotAnInteger:
+        questions = paginator.page(1)
+    except EmptyPage:
+        """Page number is out of range"""
+        questions = paginator.page(paginator.num_pages)
+
+    context = {
+        'questions': questions,
+        'email': request.user.email,
+        'categories': StationCategory.objects.all().order_by('id')
+    }
+
+    return render(request, 'app/question_list_page.html', context)
+
+
+@login_required
+def question_add_page(request):
+    if request.user.is_administrator():
+        stations = Station.objects.all()
+    else:
+        stations = Satation.objects.filter(owner_group=request.user.group)
+
+    if request.method == 'POST':
+        form = QuestionForm(request.POST)
+        if form.is_valid():
+            data = form.cleaned_data
+            question = form.save()
+
+            for i in range(1, 5):
+                choice_name = 'choice{0}'.format(i)
+                if data[choice_name]:
+                    choice = Choice.objects.get_or_create(content=data[choice_name])[0]
+                    is_answer = (choice_name == data['answer'])
+                    QuestionChoice.objects.create(question=question, choice=choice, is_answer=is_answer)
+            return HttpResponseRedirect('/questions/')
+    else:
+        form = QuestionForm()
+
+    context = {
+        'email': request.user.email,
+        'categories': StationCategory.objects.all().order_by('id'),
+        'stations': stations.order_by('id'),
+        'form': form
+    }
+    return render(request, 'app/question_add_page.html', context)
+
+
+@login_required
+def question_edit_page(request, pk):
+    question = get_object_or_404(Question, pk=pk)
+    if request.method == 'POST':
+        form = QuestionForm(request.POST, instance=question)
+        if form.is_valid():
+            data = form.cleaned_data
+            question = form.save()
+            question.choices.clear()
+            question.save()
+
+            for i in range(1, 5):
+                choice_name = 'choice{0}'.format(i)
+                if data[choice_name]:
+                    choice = Choice.objects.get_or_create(content=data[choice_name])[0]
+                    is_answer = (choice_name == data['answer'])
+                    QuestionChoice.objects.create(question=question, choice=choice, is_answer=is_answer)
```

Comment:
Line 1150~ line 1155 and line 1180~line 1185 are almost the same.
Is it possible to rewrite it as a function?

Reply:
Possible! Refactored it.